crates/ibc/src/services/core/channel.rs

```diff
@@ -0,0 +1,551 @@
+use std::boxed::Box;
+
+use ibc_proto::google::protobuf::Any;
+use ibc_proto::ibc::core::channel::v1::query_server::Query as ChannelQuery;
+use ibc_proto::ibc::core::channel::v1::{
+    PacketState, QueryChannelClientStateRequest, QueryChannelClientStateResponse,
+    QueryChannelConsensusStateRequest, QueryChannelConsensusStateResponse, QueryChannelRequest,
+    QueryChannelResponse, QueryChannelsRequest, QueryChannelsResponse,
+    QueryConnectionChannelsRequest, QueryConnectionChannelsResponse,
+    QueryNextSequenceReceiveRequest, QueryNextSequenceReceiveResponse,
+    QueryNextSequenceSendRequest, QueryNextSequenceSendResponse, QueryPacketAcknowledgementRequest,
+    QueryPacketAcknowledgementResponse, QueryPacketAcknowledgementsRequest,
+    QueryPacketAcknowledgementsResponse, QueryPacketCommitmentRequest,
+    QueryPacketCommitmentResponse, QueryPacketCommitmentsRequest, QueryPacketCommitmentsResponse,
+    QueryPacketReceiptRequest, QueryPacketReceiptResponse, QueryUnreceivedAcksRequest,
+    QueryUnreceivedAcksResponse, QueryUnreceivedPacketsRequest, QueryUnreceivedPacketsResponse,
+};
+use ibc_proto::ibc::core::client::v1::IdentifiedClientState;
+use tonic::{Request, Response, Status};
+
+use crate::core::ics04_channel::packet::Sequence;
+use crate::core::ics24_host::identifier::{ChannelId, ConnectionId, PortId};
+use crate::core::ics24_host::path::{
+    AckPath, ChannelEndPath, ClientConsensusStatePath, ClientStatePath, CommitmentPath, Path,
+    ReceiptPath, SeqRecvPath, SeqSendPath,
+};
+use crate::core::ValidationContext;
+use crate::prelude::*;
+use crate::services::core::context::QueryContext;
+use crate::Height;
+
+// TODO(rano): currently the services don't support pagination, so we return all the results.
+
+pub struct ChannelQueryService<I>
+where
+    I: QueryContext + Send + Sync + 'static,
+    <I as ValidationContext>::AnyClientState: Into<Any>,
+    <I as ValidationContext>::AnyConsensusState: Into<Any>,
+{
+    ibc_context: I,
+}
+
+impl<I> ChannelQueryService<I>
+where
+    I: QueryContext + Send + Sync + 'static,
+    <I as ValidationContext>::AnyClientState: Into<Any>,
+    <I as ValidationContext>::AnyConsensusState: Into<Any>,
+{
+    /// `ibc_context` must be thread-safe. Possibly wrapped in `Arc<Mutex<_>>` or `Arc<RwLock<_>>` or similar.
```

Comment:
"thread safety" is not the issue here; all (safe) Rust code is guaranteed to be thread safe (this is what the `Send + Sync + 'static` constraints do). I would instead write:

> `I` must be a type where writes from one thread are readable from another. This means using `Arc<Mutex<_>>` or `Arc<RwLock<_>>` in most cases.

Can you also that the same docstring to `ClientQueryService` and `ConnectionQueryService`?

Reply:
Hmm. You're right. Although this is what I meant, I see why my comment is not the same as what I meant. I will update it.